docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,571 @@
+# C++ code style guidelines
```

Comment:
I have always had it in the back of my mind to rewrite these style guidelines, and now might be the time to do it (?).

To me, they are far too long due to mentioning style choices that are obvious, such as "use const", "use smart pointers"... I think that a document that outlines the ways our style guide differs from others (trailing underscore class variables), or style that developers find contentious (how much to use templates and comments) is more valuable.

Reply:
I have removed the parts that were closer to a tutorial than style guidelines. The document should be a lot shorter now.

I have preserved those 2 notes in particular, not because I think they are unexpected, but because it is good to clarify to newcomers that we are exhaustive about `const` for instance. They are also good introductory points for other things like how do we prefer `const` declarations to look like, particulars about smart pointer handling, etc.